Move strip tracker partition and ModuleGeometry enums to TrackerTopology.h

pieterdavid opened the pull request:
#### PR description:

This implements the solution proposed in https://github.com/cms-sw/cmssw/issues/26776#issuecomment-492536913 to solve the cyclic dependency between TrackerTopology.h and SiStripDetId.h/StripSubstructure.h. I changed to the TrackerTopology version in the uses that picked up SiStripDetId.h or StripSubdetector.h through TrackerTopology.h (since the latter is preferred), but added a SiStripDetId include for the few nontrivial ones (except one where the code could easily be simplified, https://github.com/cms-sw/cmssw/compare/master...pieterdavid:removecyclictrackercommonsistripdetid?expand=1#diff-efb246c04e3568bf8622531fd4ea48d9).
The other uses of these enums should also be changed (and use of SiStripDetId reduced in favour of TrackerTopology, where possible). I could add that here, but maybe it is better to leave that for a separate PR, after all code formatting changes have landed?

#### PR validation:

Compiled on top of CMSSW_10_6_0 before rebasing to master (to solve a conflict with formatting changes in DQM/SiStripMonitorTrack/src/SiStripMonitorTrack.cc).

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26804/9811

- This PR adds an extra 472KB to repository

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: d674e9b9e8cf636192943a5ec5f56b8e6954b7f0
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-726ebd/266/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/07c6317816c06c771449ac447d2ad53df8af7253
https://github.com/cms-sw/cmssw/commit/562313fc98446faefcef05e76d4dc3cf5eb500a5
https://github.com/cms-sw/cmssw/commit/bf69e0694fa4789f3560b35cb7439413cc135058
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26804/266/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26804/266/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-726ebd/266/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 991 differences found in the comparisons
 - DQMHistoTests: Total files compared: 33
 - DQMHistoTests: Total histograms compared: 3212004
 - DQMHistoTests: Total failures: 2
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3211668
 - DQMHistoTests: Total skipped: 334
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 32 files compared)
 - Checked 137 log files, 14 edm output root files, 33 DQM output files

mmusich commented on Alignment/CommonAlignmentMonitor/plugins/AlignmentStats.cc:
```diff
@@ -163,7 +163,7 @@ void AlignmentStats::analyze(const edm::Event &iEvent, const edm::EventSetup &iS
 
       bool hitInPixel= (subDet==PixelSubdetector::PixelBarrel)||(subDet==PixelSubdetector::PixelEndcap);
-      bool hitInStrip=(subDet==SiStripDetId::TIB) || (subDet==SiStripDetId::TID) ||(subDet==SiStripDetId::TOB) ||(subDet==SiStripDetId::TEC);
+      bool hitInStrip=(subDet==TrackerTopology::TIB) || (subDet==TrackerTopology::TID) ||(subDet==TrackerTopology::TOB) ||(subDet==TrackerTopology::TEC);
```
matter of personal preference, but now that I see it done,  I find it odd to have:
`TrackerTopology::TIB  but `PixelSubdetector::PixelBarrel` in the same file

pieterdavid replied:
I also defined `TrackerTopology::PXB` and `TrackerTopology::PXF`, with the same values as in `PixelSubdetector` in case (these are 1 and 2, while the strip partitions are 3, 4, 5, and 6), so it would be trivial to change them, but I tried to keep this PR minimal (other opinions on naming etc. are welcome - I don't know how things are called for phase2, while TrackerTopology should work for all detectors).

mmusich commented on L1Trigger/TrackTrigger/plugins/TTStubBuilder.cc:
```diff
@@ -41,7 +41,7 @@ void TTStubBuilder< Ref_Phase2TrackerDigi_ >::produce( edm::Event& iEvent, const
     DetId detid = (*gd).geographicalId();
-    if(detid.subdetId()!=StripSubdetector::TOB && detid.subdetId()!=StripSubdetector::TID ) continue; // only run on OT
+    if(detid.subdetId()!=TrackerTopology::TOB && detid.subdetId()!=TrackerTopology::TID ) continue; // only run on OT
     if(!tTopo->isLower(detid) ) continue; // loop on the stacks: choose the lower arbitrarily
```
not related to the purpose of this PR, but why do we even have TOB and TIB in phase-2 code?

pieterdavid replied:
I do not know if this is correct (especially since the TrackerGeometryBuilder documentation that @venturia pointed to says that Phase2 only has 4 and 5), but https://github.com/cms-sw/cmssw/blob/master/Geometry/TrackerCommonData/data/PhaseII/trackerParameters.xml defines a scheme for all subdetectors 3, 4, 5 and 6, so at least for the purpose of DetIds (as interpreted by the TrackerTopology) they are defined

venturia replied:
Whoever modified the code since when I wrote the documentation should have updated the README file consistently. Otherwise we oscillate between no documentation and out-of-date documentation

venturia commented:
@mmusich all this business of the enumerators to identify the Tracker subdetector is pretty complicated. When I did the work to merge present and phase-2 code several years ago I tried to clean (but not drastically) and to describe (maybe not thoroughly) the situation in the README files of the geometry packages (for example here https://github.com/venturia/cmssw/tree/CMSSW_7_4_X/Geometry/TrackerGeometryBuilder )
The issue can be summarized in this way: in some parts of the code there is the need to have something which counts from 1 to 6, where 6 is the largest number of subdetectors that the Tracker can have for a given geometry (there are several vectors and arrays that have 6 as hardcoded size and in some places the code relies on that). In other parts of the code, instead, having one enumerator for each possible Tracker subdetectors, combinging together phase-0, phase-1, phase-2, other exotic geomtries,... is needed. This leads, in some cases to the ambiguity of using TIB and TOB also for phase-2 in some cases. I am available to explain what I did on my side to anyone interested to continue the clean up

slava77 commented:
@pieterdavid 
 note that this PR does not merge anymore.
Please rebase/update.
Thank you.

slava77 commented:
If I'm not mistaken a more minimal change would be to 
- introduce DataFormats/TrackerCommon/interface/SiStripEnums.h with only header content (actually, with just the header dependency, it can even stay in DataFormats/SiStripDetId)
    - ```SiStripSubDetector { UNKNOWN = 0, TIB = 3, TID = 4, TOB = 5, TEC = 6 }```
    - ```SiStripModuleGeometry { UNKNOWNGEOMETRY, IB1, IB2, OB1, OB2, W1A, W2A, W3A, W1B, W2B, W3B, W4, W5, W6, W7 }```
- use this instead of ```SiStripDetId::SubDetector``` or ```StripSubdetector::SubDetector``` and ```SiStripDetId::ModuleGeometry``` in the DataFormats/TrackerCommon/interface/TrackerTopology.h 
- perhaps also go into DataFormats/SiStripDetId/interface/SiStripDetId.h and DataFormats/SiStripDetId/interface/StripSubdetector.h and replace the values defined there with those from DataFormats/TrackerCommon/interface/SiStripEnums.h
    - e.g. in StripSubdetector ```using SubDetector = SiStripSubDetector;```

This would avoid a need to parse the full TrackerTopology.h for all use cases where just the enums are needed.

pieterdavid committed:
Move SiStripDetId subdetector and module geometry enums

to new header DataFormats/TrackerCommon/interface/SiStripEnums.h;
SiStripDetId::SubDetector and StripSubdetector::SubDetector defined in
terms of the new SiStripSubdetector, and SiStripDetId::ModuleGeometry
changed to SiStripModuleGeometry

pieterdavid committed:
Add missing includes

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26804/10012

- This PR adds an extra 476KB to repository

- Found files with invalid states:
  - PhysicsTools/TensorFlow/test/test_graph_tfadd.pb:
    - Deleted: 4de068bc620c5d70b8b1bd56caf0f86aa9aeae08
    - Added: 0b9536e2bbf1ef030c9134ecfebc326fbe5583ed
    - Modified: 43e84aac1e58fee67927b1b8c5d464d6d112ebcd

pieterdavid commented:
Sorry, I rebased on the wrong branch (DEVEL IB instead of regular IB), I'm fixing it...

cmsbuild commented:
Pull request #26804 was updated. @andrius-k, @lveldere, @schneiml, @ianna, @kpedro88, @rekovic, @fioriNTU, @tlampen, @pohsun, @santocch, @perrotta, @civanch, @cmsbuild, @smuzaffar, @Dr15Jones, @cvuosalo, @mdhildreth, @jfernan2, @tocheng, @slava77, @ggovi, @ssekmen, @kmaeshima, @christopheralanwest, @franzoni can you please check and sign again.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26804/10014

- This PR adds an extra 472KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-checks**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26804/10014/code-checks.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26804/10014/code-checks.patch | patch -p1`
You can also run `scram build code-checks` to apply code checks directly

pieterdavid committed:
code-checks fixes

pieterdavid commented:
This version implements the suggestion by @slava77 in https://github.com/cms-sw/cmssw/pull/26804#issuecomment-494175044 .
For the files that relied on one header including the other I added the include for now; it is possible to replace SiStripDetId::SubDetector and StripSubdetector::SubDetector by SiStripSubdetector::Subdetector (that's a search-and-replace exercise involving quite some files) - let me know if you prefer that I add it here instead of in a followup PR.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26804/10015

- This PR adds an extra 304KB to repository

cmsbuild commented:
Pull request #26804 was updated. @andrius-k, @lveldere, @schneiml, @ianna, @kpedro88, @rekovic, @fioriNTU, @tlampen, @pohsun, @perrotta, @civanch, @cmsbuild, @Dr15Jones, @cvuosalo, @mdhildreth, @jfernan2, @tocheng, @slava77, @ggovi, @ssekmen, @kmaeshima, @christopheralanwest, @franzoni can you please check and sign again.

slava77 commented:
> let me know if you prefer that I add it here instead of in a followup PR.

The full update will probably touch many files.
I think that such a change better be done after the clang-format campaign is over.
For now let's focus on the minimal part to resolve  #26776

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: 76a87c52954de2da3f00ccc9963df8a8adf6d0af

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-726ebd/544/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests:

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-726ebd/544/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 33
 - DQMHistoTests: Total histograms compared: 3209232
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3208897
 - DQMHistoTests: Total skipped: 334
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 32 files compared)
 - Checked 137 log files, 14 edm output root files, 33 DQM output files

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/26804	76a87c5
- code changes are in line with the PR description and the follow up review
    - some of the changes are perhaps superfluous, but I think that they are fine
- jenkins tests pass and comparisons with the baseline show

kpedro88 commented:
@pieterdavid there are merge conflicts, can you rebase?

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26804/10105

- This PR adds an extra 316KB to repository

cmsbuild commented:
Pull request #26804 was updated. @andrius-k, @lveldere, @schneiml, @ianna, @kpedro88, @rekovic, @fioriNTU, @tlampen, @pohsun, @perrotta, @civanch, @cmsbuild, @Dr15Jones, @cvuosalo, @mdhildreth, @jfernan2, @tocheng, @slava77, @ggovi, @ssekmen, @kmaeshima, @christopheralanwest, @franzoni can you please check and sign again.

jfernan2 commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: 5a8f1f8bf3b98697cc71a3336a9f037e1b88e35b

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-726ebd/644/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals**

* **RelVals**:

When I ran the RelVals I found an error in the following workflows:
27434.0 step2
<pre>runTheMatrix-results/27434.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D35_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D35+RecoFullGlobal_2023D35+HARVESTFullGlobal_2023D35/step2_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D35_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D35+RecoFullGlobal_2023D35+HARVESTFullGlobal_2023D35.log

</pre>29034.0 step2
<pre>runTheMatrix-results/29034.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D41_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D41+RecoFullGlobal_2023D41+HARVESTFullGlobal_2023D41/step2_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D41_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D41+RecoFullGlobal_2023D41+HARVESTFullGlobal_2023D41.log

</pre>

cmsbuild commented:
Comparison not run due to runTheMatrix errors (RelVals and Igprof tests were also skipped)

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 5a8f1f8bf3b98697cc71a3336a9f037e1b88e35b
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-726ebd/676/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-726ebd/676/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 2 differences found in the comparisons
 - DQMHistoTests: Total files compared: 33
 - DQMHistoTests: Total histograms compared: 3210678
 - DQMHistoTests: Total failures: 2
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3210342
 - DQMHistoTests: Total skipped: 334
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 32 files compared)
 - Checked 137 log files, 14 edm output root files, 33 DQM output files

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/26804	5a8f1f8
- changes since the last reco signoff (https://github.com/cms-sw/cmssw/pull/26804	76a87c5) are related to the rebase 
- jenkins tests pass and comparisons with the baseline show no differences

fabiocos commented:
@civanch @ggovi @ssekmen @cvuosalo @kpedro88 @andrius-k @rekovic could you please have a look?

jfernan2 commented:
+1

ssekmen commented:
+1

cvuosalo commented:
+1

slava77 commented:
alca, DB, upgrade, and L1 signatures are still missing.

@fabiocos  perhaps you can check this at some point soon to possibly override the missing signatures.
Thank you.

christopheralanwest commented:
+1

kpedro88 commented:
+upgrade

fabiocos commented:
@rekovic @ggovi @civanch the changes in your areas look simple to me, please check, I would like to move forward with this PR by today

civanch commented:
+1

fabiocos commented:
+1

fabiocos commented:
merge
